### PR TITLE
feat(bucket): support for realtime changes

### PR DESCRIPTION
**Pull request by thesayyn**

### CODE REVIEW

File: packages/database/realtime/database.service.ts [hunk 1]
```diff
@@ -0,0 +1,280 @@
...
+            filter(changes => changes.length > 0)
+          )
+          .subscribe(async events => {
+            let syncedIds: string[];
+
+            if (
+              sortedKeys.length == 1 &&
+              sortedKeys[0] == "_id" &&
+              events.length > 1 &&
+              options.sort._id == -1
```
Comment (emre2345) [inline packages/database/realtime/database.service.ts]: @thesayyn What does this if block mean?

**Comment by emre2345** [comment]
@thesayyn 

Do you think we might add more items into the FilterOptions?

```
export interface FindOptions<T> {
  filter?: FilterQuery<T>;
  sort?: {
    [index: string]: -1 | 1;
  };
  skip?: number;
  limit?: number;
}
```

**Comment by emre2345** [comment]
@thesayyn We should split the `find` function in the packages/database/realtime/database.service.ts

File: packages/database/realtime/database.service.ts (hunk 1 above)
Comment (thesayyn) [reply packages/database/realtime/database.service.ts]: This check should be removed as well as the other commented lines. 

My initial intention was to make the `_id` sort faster by sorting the _id directly in the process.

**Comment by thesayyn** [comment]
Why?

**Comment by emre2345** [comment]
> Why?

Because it's too long and have too many cascade operations. This makes it hard to understand what's going on in the function.

Also need to go through the logic implemented to find a way to make it more understandable. For instance `getMore` variable doesn't make sense to me.

File: packages/database/realtime/database.service.ts (hunk 1 above)
Comment (emre2345) [reply packages/database/realtime/database.service.ts]: Ok so let's remove them then.

**Comment by thesayyn** [comment]
In order to do that, we have to simplify the logic first. Also, if you are similar to the syntax, it is not a big deal IMHO.

**Comment by emre2345** [comment]
Generally speaking having a function which has around 270 lines isn't something to do with being accustomed to. This is a rule of thumb having small functions. This function violates single responsibility rule which is one of the most important rules in the software development.

I created a tech story https://dreamharvesters.atlassian.net/browse/SP-140

Let's leave this as it's right now and we look into this later once we complete the urgent feature requests.

**Comment by emre2345** [comment]
@thesayyn Went through all the codes and everything seems fine.

File: packages/database/realtime/database.service.ts [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+    const changedKeys: string[] = (change.updateDescription.removedFields || []).concat(
+      Object.keys(change.updateDescription.updatedFields || {})
+    );
+    return changedKeys.some(k => sortedKeys.indexOf(k) > -1);
+  }
+
+  return true;
+}
+
+function fetchSortedIdsOfTheCursor<T>(
```
Comment (emre2345) [inline packages/database/realtime/database.service.ts]: @thesayyn Do you think we might need more if checks in this function?

File: packages/database/realtime/database.service.ts [hunk 3]
```diff
@@ -0,0 +1,312 @@
...
+function reverseFilter<T>(filter: FilterQuery<T>) {
+  return Object.keys(filter).map(key => {
+    const value = filter[key];
+    return {
+      [`fullDocument.${key}`]: typeof value == "object" ? {$not: value} : {$ne: value}
+    };
+  });
+}
+
+function getSortedKeys(sort: {[k: string]: -1 | 1}) {
```
Comment (emre2345) [inline packages/database/realtime/database.service.ts]: @thesayyn Can you please explain how keys are sorted with this function?

**Comment by emre2345** [comment]
@thesayyn The last commit makes things much cleaner. Good work.

I kind of understand the approach in the find method. You setup sort, fetchMoreToFillCursor and some other functions and use them later in the find logic. 

I wonder if it would be better moving the function setups to outside of the find function. I felt like all the codes in the find method was the continuous flow of the intended logic.

File: packages/database/realtime/database.service.ts [hunk 4]
```diff
@@ -0,0 +1,312 @@
...
+                fullDocument: "updateLookup"
+              }
+            )
+            .on("change", change => {
+              if (hasChangeAlreadyPresentInCursor(ids, change)) {
+                observer.next({
+                  kind: ChunkKind.Expunge,
+                  document: change.documentKey
+                });
+                ids.delete(change.documentKey._id.toString());
```
[the remark this answers is not in the document]
Comment (thesayyn) [reply packages/database/realtime/database.service.ts]: "toHexString" method is the alias of the "toString" method. Good catch I should make them use the same methods.

File: packages/database/realtime/database.service.ts (hunk 3 above)
Comment (thesayyn) [reply packages/database/realtime/database.service.ts]: As a matter of fact, this function just returns the sorted keys of the cursor. 
Shall I rename it to `getSortedKeysOfTheCursor`? Would it make it more understandable?

File: packages/database/realtime/database.service.ts (hunk 2 above)
Comment (thesayyn) [reply packages/database/realtime/database.service.ts]: For now, I don't think that we'll need more "if statements" in this function.

File: packages/database/realtime/database.service.ts (hunk 3 above)
Comment (emre2345) [reply packages/database/realtime/database.service.ts]: @thesayyn I don't get what "sorted" means. In what manner do we sort the keys and why do we sort them? :)

Comment (thesayyn) [reply packages/database/realtime/database.service.ts]: In order to sort the cursor, the client provides the name(s) of the properties that he/she want to sort. 

A complete sorting condition looks like this
`{test:-1, anotherprop:1}`.

So this method returns only the keys of the sorting as an array.

**Comment by thesayyn** [comment]
@emre2345 This PR was the blocker story of the https://dreamharvesters.atlassian.net/browse/SP-140 tech story. I think we have accomplished what we wanted. 

Could you please move SP-140 to Done if you also think this tech story is completed?

**Pull request merged**